### PR TITLE
Regex support for catching rules

### PR DESCRIPTION
**Pull request by un-tone**
For #54 
I'm considering this feature as experimental so I have not changed the data model. I just use prefix `\regex\` to mark regex rules. For example, use `\regex\\/WebsiteEditor\/surface\/(\w|\d)*\/general` catch URI to apply catching rule for general pages of all surfaces.

Currently we have the following priorities for catch URI matching:

1. Exact match
2. Match with wildcard (described in README)
3. Match with regex
4. Catch-all

image

This PR should not close the #54 issue since we should make additional improvements.
Here I'd want to get a feedback on my chosen approach.

### CODE REVIEW

**Comment by PatrykSzwer** [comment]
I think, speaking about UI itself, It should be some indicator like checkbox, to say that "This URI is with REGEX", not a part or URI itself. It would be more intuitive and readable.

**Comment by un-tone** [comment]
We should consider these #191 changes when doing it.

**Comment by warpech** [comment]
I agree with @PatrykSzwer regarding the checkbox (like in Notepad++)

**Comment by un-tone** [comment]
Please review the PR. I've updated UI and added new test.

**Comment by un-tone** [comment]
Requires https://github.com/StarcounterApps/Simplified/pull/75

File: src/WebsiteEditor/wwwroot/WebsiteEditor/viewmodels/CatchingRulesPage.html [hunk 1]
```diff
@@ -37,11 +37,15 @@
                     </div>
                     <div class="website-paper-collapse-item-text-input-section">
                         <label for="websiteCatchingRuleUri">Catch URI: </label>
-                        <input type="text" id="websiteCatchingRuleUri" value="{{catchingRule.Url$::change}}" class="website-paper-collapse-item-text-input-section__input"/>
+                        <input type="text" id="websiteCatchingRuleUri" value="{{catchingRule.Url$::change}}" class="website-paper-collapse-item-text-input-section__input" />
+                    </div>
+                    <div>
+                        <label for="websiteCatchingRuleIsRegex">Regular expression format</label>
```
Comment (warpech) [inline src/WebsiteEditor/wwwroot/WebsiteEditor/viewmodels/CatchingRulesPage.html]: Can we change this label to one of the below:

In Atom it is called "Use Regex", in VS Code it is called "Use Regular Expression", in VS it is called "Use Regular Expressions".

**Comment by PatrykSzwer** [comment]
I've managed to crash website with regex/uri like this:
`/[a-zA-Z]+/[a-zA-Z]+`

File: src/WebsiteEditor/wwwroot/WebsiteEditor/viewmodels/CatchingRulesPage.html (hunk 1 above)
Comment (un-tone) [reply src/WebsiteEditor/wwwroot/WebsiteEditor/viewmodels/CatchingRulesPage.html]: Changed to "Use Regular Expression"

**Comment by un-tone** [comment]
@PatrykSzwer, I guess you are talking about infinite loop when used nestable rule. I fixed it.

**Comment by PatrykSzwer** [comment]
I'm not sure if It's expected behavior. Right now when I'm using regex that I pointed before, my result look like this:
image
I think It's not correct. My regex says, that URL for route has to be `/something/something` to be applied. 
`/[a-zA-Z]+/[a-zA-Z]`. But It's applied even when URI is: `/WebsiteEditor/surface/DyC/catchingrules` so `/something/something/something/something`. Except that, also this sidebar, created 2 times doesn't look good. 

It's somehow caused because of `+` sign. When I remove + everything works as expected.

**Comment by un-tone** [comment]
It seems that your regex is incorrect and invalid. For `/something/something` route the correct regex should be `\/[a-zA-Z]+\/[a-zA-Z]+`

**Comment by PatrykSzwer** [comment]
Hmm In that case maybe we should do something about incorrect regex? I'm not sure tho how could we do It, since probably typical checking with `try...catch` wouldn't throw an exception. Do You have any idea? Or should we just left it as it is?

**Comment by un-tone** [comment]
I think we should left it as is. Otherwise we have to impose restrictions on regular expressions. I guess we don't want to do it.

TBH, I can't reproduce your case. @PatrykSzwer could you provide more detailed steps for it?

File: src/WebsiteProvider/Api/ContentHandlers.cs [hunk 2]
```diff
@@ -181,6 +173,19 @@ protected SurfacePage GetSurfacePage(WebTemplate template)
             return page;
         }
 
+        protected string GetWildCardUrl(string url)
+        {
+            var reg = new Regex(@"[/]\w*$", RegexOptions.IgnoreCase);
```
Comment (Mihaiii) [inline src/WebsiteProvider/Api/ContentHandlers.cs]: Why only `\w`? Are we sure that we can't get a url that has, for example, "-"?

Comment (un-tone) [reply src/WebsiteProvider/Api/ContentHandlers.cs]: Since we are replacing this feature in #191 and here, it doesn't matter anymore 😄

**Comment by Mihaiii** [comment]
But I also agree that the user should be notified if he uses a wrong regex. I'm not aiming for a job done 100% right as we are talking about different engines, but you could validate it in JS and display a message like "The regex **might** be wrong". Maybe even be sure that it displays a slightly different message for (negative) lookbehind (as here might be the a lot of cases in which the notification is wrong) ?

I'm thinking that it might lead to real problems for clients if they are sure that they entered a valid regex and after a while it doesn't work as they expected and we all start digging around to see why it doesn't work. :)

**Comment by PatrykSzwer** [comment]
@un-tone to reproduce my case, just go to sidebar/responsive surface and in catching rules make catching rule with catch url: `/[a-zA-Z]+/[a-zA-Z]`
image

**Comment by un-tone** [comment]
@Mihaiii @PatrykSzwer such validation of regular expressions is a complex issue and I don't have a simple/fast solution for that. I guess currently nobody has. It means that should be implemented separately if we decide to do it.
Maybe we could create a test page for checking URLs and corresponding routes.

**Comment by un-tone** [comment]
@PatrykSzwer I still can't reproduce the behaviour of your case. On my computer it falls the app down.

**Comment by warpech** [comment]
We cannot detect incorrect regexpes but at least we can detect unparseable regexp, maybe let's do that?

**Comment by un-tone** [comment]
@PatrykSzwer I fixed that bag `/[a-zA-Z]+/[a-zA-Z]`, please check it.

@warpech it sounds good, but regex validation with Regex doesn't suite such cases. So I didn't add such detecting when editing a route URI, also since the app doesn't have a proper user notification mechanism. But I added exception catching in WebsiteProvider middleware when regex creating and matching.

**Comment by PatrykSzwer** [comment]
The bug is fixed, good job! 
I'm still not sure though if It's working correctly. I used this regex for test: `/[a-zA-Z]+e/[a-zA-Z]`
As I tested on `https://regexr.com/` this regex should catch URI for example `/people/persons` but shouldn't catch uri like `http://localhost:8080/WebsiteEditor/surface/Bs/catchingrules`. Unfortunately both are catched. But to make it more interesting,  regex like: `/[a-zA-Z]+e/[a-zA-Z]+s` is catched in `/people/persons` but not in any other not matching URI, except sidebarSurface page.

**Comment by un-tone** [comment]
Why don't you escape `/` symbols? I think it should be `\/[a-zA-Z]+e\/[a-zA-Z]`

**Comment by PatrykSzwer** [comment]
For some reason I though in our case we don't have to escape `/` symbol because We're parsing it and cuting URI. 
So one more thing now, when I escaped `/` symbols. Regex: `/[a-zA-Z]+e\/[a-zA-Z]+` should catch `/people/persons` but shouldn't `/WebsiteEditor/surface/Bs/catchingrules` (confirmed with `regexr.com`). Although It's still catched in both cases.

**Comment by un-tone** [comment]
> I guess It can be catched by mistake because surface/Bs part is matched.

Not by mistake.
You should use `^(\/[a-zA-Z]+e\/[a-zA-Z]+)` expression for your case.

**Pull request merged**